macros/weightedGrader.pl

```diff
@@ -0,0 +1,330 @@
+  } else {
+    my $newChecker = sub {
+      my $hash = &{$checker}(@_);
+      $hash->{weight} = $weight;
+      $checker->{rh_ans}->{credit} = $credit if defined($credit);
+      return $hash;
+    };
+    return $newChecker;
+  }
+}
+
+
+##################################################
+#
+#  This is the weighted grader.  It uses an extra field added to the
+#  AnswerHash (named "weight") to tell how much weight to give each
+#  problem.  The grader adds up the total weights for the correct
+#  answers.  For partially correct ones, it uses the score for that
+#  answer to give a portion of that weight.  For example, if the
+#  weight is 40 and the score for the answer is .5, then 20 is added
+#  to the total for the problem.  (Note that most answer checkers only
+#  return 1 or 0, but they are allowed to return partial credit as
+#  well.)
+#
+#  When the student's total is computed, it is divided by the sum of
+#  all the weights in order to determine the final score. 
+#
+#  It also uses a special field called "credit" that determines
+#  what other (named) answers are given full credit if the given
+#  answer is correct.  This can be used to make "optional" answers,
+#  where getting the "final" answer right gives credit for the other parts.
+#
+
+sub weighted_grader {
+  my $rh_evaluated_answers = shift;
+  my $rh_problem_state = shift;
+  my %form_options = @_;
+  my %answers = %{$rh_evaluated_answers};
+  my %problem_state =	%{$rh_problem_state};
+    
+  my %problem_result = (
+    score  => 0,
+    errors => '',
+    type   => 'weighted_grader',
+    msg    => '',
+  );
+    
+  if (scalar(keys(%answers)) == 0) {
+    $problem_result{msg} = "This problem did not ask any questions.";
+    return(\%problem_result,\%problem_state);
+  }
+    
+  return(\%problem_result,\%problem_state)
+    if (!$form_options{answers_submitted});
+    
+  my ($score,$total) = (0,0);
+  my ($weight,$ans_name,$credit_name);
+  my (%credit);
+
+  #
+  #  Get the score for each answer
+  #  (error if can't recognize the answer format).
+  #
+  foreach $ans_name (keys %answers) {
+    if (ref($answers{$ans_name}) =~ m/^(HASH|AnswerHash)$/) {
+      $credit{$ans_name} = $answers{$ans_name}->{score};
+    } else {
+      die "Error at file ",__FILE__,"line ", __LINE__,": Answer |$ans_name| " .
+	"is not a hash reference\n" . $answers{$ans_name} .
+	"\nThis probably means that the answer evaluator for ".
+	"this answer is not working correctly.";
+      $problem_result{error} =
+	"Error: Answer $ans_name is not a hash: $answers{$ans_name}";
+    }
+  }
+
+  #
+  #  Mark any optional answers as correct, if the goal answers
+  #  are right and the optional ones are blank.
+  #
+  foreach $ans_name (keys %answers) {
+    if ($credit{$ans_name} == 1 &&
+	defined($answers{$ans_name}->{credit})) {
+      foreach $credit_name (@{$answers{$ans_name}->{credit}}) {
+	$credit{$credit_name} = 1 
+	  if ( ($answers{$credit_name}->{student_ans} =~ s/^\s+|\s+$//gr) eq "");
```

Comment:
This is what I initially put into my comment and works, but I realized that this is a bit silly.  Rather than using a substitution to trim whitespace from the beginning and end of the string and then comparing to the empty string, just use a match against /^\s*$/.  I updated my comment with this.

Reply:
Thanks. The first code was also missing a parenthesis... and after pushing I noticed your change. I force pushed a file with the revised change.